apps/api/src/http/internal/rest/webhooks/stripe/checkout_session.rs

```diff
@@ -0,0 +1,253 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use shared::database::product::codes::{RedeemCode, RedeemCodeId};
+use shared::database::product::subscription::{
+	ProviderSubscriptionId, SubscriptionId, SubscriptionPeriod, SubscriptionPeriodCreatedBy, SubscriptionPeriodId,
+};
+use shared::database::product::SubscriptionProductId;
+use shared::database::queries::{filter, update};
+use shared::database::user::UserId;
+
+use crate::global::Global;
+use crate::http::error::ApiError;
+use crate::http::v3::rest::egvault::redeem::grant_entitlements;
+use crate::transactions::{TransactionError, TransactionResult, TransactionSession};
+
+pub async fn completed(
+	global: &Arc<Global>,
+	mut tx: TransactionSession<'_, ApiError>,
+	session: stripe::CheckoutSession,
+) -> TransactionResult<Option<SubscriptionId>, ApiError> {
+	let Some(metadata) = session.metadata else {
+		// ignore metadata sessions that don't have metadata
+		return Ok(None);
+	};
+
+	if session.mode == stripe::CheckoutSessionMode::Setup {
+		// setup session
+		// the customer successfully setup a new payment method, now set it as the
+		// default payment method
+
+		let setup_intent = session
+			.setup_intent
+			.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?
+			.id();
+		let setup_intent = stripe::SetupIntent::retrieve(&global.stripe_client, &setup_intent, &[])
+			.await
+			.map_err(|e| {
+				tracing::error!(error = %e, "failed to retrieve setup intent");
+				TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+			})?;
+
+		let customer_id = session.customer.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?.id();
+		let Some(payment_method) = setup_intent.payment_method.map(|p| p.id()) else {
+			return Ok(None);
+		};
+
+		let customer = stripe::Customer::update(
+			&global.stripe_client,
+			&customer_id,
+			stripe::UpdateCustomer {
+				invoice_settings: Some(stripe::CustomerInvoiceSettings {
+					default_payment_method: Some(payment_method.to_string()),
+					..Default::default()
+				}),
+				..Default::default()
+			},
+		)
+		.await
+		.map_err(|e| {
+			tracing::error!(error = %e, "failed to update customer");
+			TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+		})?;
```

Comment:
We should have some global limit for the number of api calls we can do concurrently to stripe

Reply:
We should, but do we care?

Reply:
Yeah, probs. You can use a semaphore.